lib/analyser.py

```diff
@@ -1,11 +1,206 @@
+                else:
+                    margin_score = 4
+            elif(decrease_index_year == 3):
+                if(ebit_margins[decrease_index_year-1] <= ebit_margins[decrease_index_year+1]):
+                    margin_score = 5
+                elif(ebit_margins[decrease_index_year-1] <= ebit_margins[decrease_index_year+2]):
+                    margin_score = 4
+            elif(decrease_index_year == 2):
+                if(ebit_margins[decrease_index_year-1] <= ebit_margins[decrease_index_year+1]):
+                    margin_score = 5
+                elif((ebit_margins[decrease_index_year-1] <= ebit_margins[decrease_index_year+2]) or (ebit_margins[decrease_index_year-1] <= ebit_margins[decrease_index_year+3])):
+                    margin_score = 4
+                else:
+                    margin_score = 3
+
+        elif((yoy_margin_increase_count <= 3) or (1 <= yoy_margin_increase_count)):
+            if(median_ebit_margin >= 12):
+                if(max_min_ebit_margin <= 4):
+                    margin_score = 4
+                elif(max_min_ebit_margin <= 6):
+                    margin_score = 3
+                else:
+                    margin_score = 2
+            elif(median_ebit_margin >= 8):
+                if(max_min_ebit_margin <= 4):
+                    margin_score = 2
+                else:
+                    margin_score = 1
+            else:
+                margin_score = 0
+        else:
+            margin_score = 0
+
+    stock.yoy_margin_increase_count = yoy_margin_increase_count
+    stock.negative_margin_count = negative_margin_count
+    stock.margin_score = margin_score
+
+def dividend_check(stock: Stock):
+    """
+    dpr should be under 20% for the last 5 years
+    dividend against debt check
+    """
+    # The last 5 years dpr
+    dpr_raw = list(stock.financial_data_raw.loc['dpr'])[-5:]
+    dpr_clean = np.array([int(dpr.split("%")[0]) for dpr in dpr_raw])
+    dpr_mean = dpr_clean.mean()
+
+    borrowings = (stock.balance_sheet_data_raw.loc['borrowings']).astype(np.float)[-5:]
+    net_profit = (stock.financial_data_raw.loc['net_profit']).astype(np.float)[-5:]
+
+    if(dpr_mean <= 20):
+        stock.controlled_dpr = True
+    else:
+        stock.controlled_dpr = False
+
+    stock.idiot_dividend_policy = (borrowings > net_profit) & (dpr_clean > 10)
+    stock.idiot_dividend_policy_points = len(np.nonzero(idiot_management_arr))
+
+def roa_efficiency(stock: Stock):
+    """
+    If the RoA is increasing while the Dividend payout is under 20%, then the company is improving 
+    asset utilization -> come up with equation where for given year dividend payout if not given
+    then what is the RoA and then see if it is better than last year's actual RoA
+    """
+    
+
+
+# def margin_of_safety(stock: Stock):
+
+
+# def debt_check(stock: Stock):
+    """
+    compare debt against profits
+    compare debt against capex and cwip
+    """
+
+
+# def growth_check(stock: Stock):
+
+
+
+
+
```

Comment:
No point in checking in commented code and a bunch of newlines
```suggestion
"""
TODO:
  - margin_of_safety
  - debt_check: compare against profits, capex and cwip
  - growth_check
"""
```